### PR TITLE
fix(ui): `Timeline::send_reply` correctly sets up `m.mentions`

### LINKED ISSUE #3314: When user reply to their message, the reply mentions the current user
Originally opened on Ruma project: https://github.com/ruma/ruma/issues/1770 (see details there), but the fix seems to be located in the Rust project.

### PR DESCRIPTION
**Pull request by Hywan**
In https://github.com/matrix-org/matrix-rust-sdk/pull/2691, I suppose
the way `add_mentions` is computed is… wrong. [`AddMentions`](https://docs.rs/ruma/latest/ruma/events/room/message/enum.AddMentions.html) is used to
automatically infer the [`m.mentions`](https://spec.matrix.org/v1.10/client-server-api/#definition-mmentions) of the reply event based on the
replied event. The way it was computed was based on the reply event
`mentions`, which seems wrong: if the reply contains mentions, then the
sender should be part of it? Nah. That's a bug. We want the reply event
to automatically mention the sender of the replied event if and only
if it's not the same as the current user, i.e. the sender of the reply
event.

This patch fixes the `add_mentions` calculation. This patch also updates
a test and adds another test to ensure that `m.mentions` is correctly
defined when replying to an event.

---

* Fix #3314 
* Address https://github.com/element-hq/element-x-android/issues/2671

### CODE REVIEW

File: crates/matrix-sdk-ui/src/timeline/mod.rs [hunk 1]
```diff
@@ -312,8 +311,24 @@ impl Timeline {
...
+        // replied event (`reply_item`)…
+        let mut mention_the_sender_of_the_replied_event = AddMentions::Yes;
+
+        // … but if `reply_item` has been written by the current user, let's toggle
+        // to `AddMentions::No`.
+        if let Some(current_user_id) = client.user_id() {
+            if current_user_id == reply_item.sender {
+                mention_the_sender_of_the_replied_event = AddMentions::No;
+            }
+        }
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/mod.rs:331]: nit: Maybe we can avoid a bit of verbosity, the `client` local variable, and the `mut` by having something like:

```suggestion
        let mention_sender = if self.room().own_user_id() == reply_item.sender { AddMentions::No } else { AddMentions::Yes };
```

File: crates/matrix-sdk-ui/src/timeline/mod.rs [hunk 2]
```diff
@@ -325,7 +340,11 @@ impl Timeline {
                     content: msg.to_content(),
                     unsigned: Default::default(),
                 };
-                content.make_reply_to(&event, forward_thread, add_mentions)
+                content.make_reply_to(
+                    &event,
+                    forward_thread,
+                    mention_the_sender_of_the_replied_event,
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/mod.rs:346]:
```suggestion
                    mention_sender,
```

File: crates/matrix-sdk-ui/src/timeline/mod.rs [hunk 3]
```diff
@@ -337,7 +356,7 @@ impl Timeline {
                     event_id.to_owned(),
                     self.room().room_id(),
                     forward_thread,
-                    add_mentions,
+                    mention_the_sender_of_the_replied_event,
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/mod.rs:359]:
```suggestion
                    mention_sender,
```

**Review by bnjbvr — APPROVED** [review]
Thanks, that's a sweet test!

File: crates/matrix-sdk-ui/src/timeline/mod.rs (hunk 1 above)
Comment (Hywan) [reply crates/matrix-sdk-ui/src/timeline/mod.rs]: Indeed, that's better, thanks!

**Pull request merged**